File: api/shared/services/creditService.ts
import { CosmosService } from './cosmosService';
import { User } from '../models/user';
import { CreditTransaction } from '../models/creditTransaction';
import { v4 as uuidv4 } from 'uuid';

export class CreditService {
  private cosmosService: CosmosService;

  constructor() {
      this.cosmosService = new CosmosService();
  }

  async getUserCredits(userId: string): Promise<number> {
      const user = await this.cosmosService.getUser(userId);
      if (!user) {
          // Create new user with initial credits
          const newUser: User = {
              id: userId,
              userId,
              email: '', // Will be updated later
              credits: 15, // Initial free credits
              createdAt: new Date().toISOString(),
              updatedAt: new Date().toISOString()
          };
          await this.cosmosService.createUser(newUser);
          return newUser.credits;
      }
      return user.credits;
  }

  async deductCredits(userId: string, amount: number, description: string): Promise<number> {
      const user = await this.cosmosService.getUser(userId);
      if (!user) throw new Error('User not found');
      if (user.credits < amount) throw new Error('Insufficient credits');

      const newBalance = user.credits - amount;
      await this.cosmosService.updateUserCredits(userId, newBalance);

      const transaction: CreditTransaction = {
          id: uuidv4(),
          userId,
          amount: -amount,
          type: 'DEDUCTION',
          description,
          createdAt: new Date().toISOString()
      };
      await this.cosmosService.createTransaction(transaction);

      return newBalance;
  }

  async addCredits(userId: string, amount: number, description: string, reference?: string): Promise<number> {
      const user = await this.cosmosService.getUser(userId);
      if (!user) throw new Error('User not found');

      const newBalance = user.credits + amount;
      await this.cosmosService.updateUserCredits(userId, newBalance);

      const transaction: CreditTransaction = {
          id: uuidv4(),
          userId,
          amount,
          type: 'PURCHASE',
          description,
          reference,
          createdAt: new Date().toISOString()
      };
      await this.cosmosService.createTransaction(transaction);

      return newBalance;
  }
}

File: api/shared/services/credit_service.py
from datetime import datetime
import uuid
import logging
from typing import List, Optional
from .cosmos_service import CosmosService
from ..models.user import User
from ..models.credit_transaction import CreditTransaction

class CreditService:
    def __init__(self):
        self.cosmos_service = CosmosService()

    async def get_user_credits(self, user_id: str) -> int:
        logging.info(f"Getting the details of user with User ID: {user_id}")
        user = await self.cosmos_service.get_user(user_id)
        if not user:
            # Create new user with initial credits
            logging.info(f"User doesn't exist. Creating new user in the DB with User ID: {user_id}")
            new_user = User(
                id=user_id,
                user_id=user_id,
                email='',  # Will be updated later
                credits=15,  # Initial free credits
                created_at=datetime.utcnow().isoformat(),
                updated_at=datetime.utcnow().isoformat()
            )
            await self.cosmos_service.create_user(new_user)
            return new_user.credits
        logging.info(f"User Credits: {user.credits}")
        return user.credits

    async def deduct_credits(self, user_id: str, amount: int, description: str) -> int:
        user = await self.cosmos_service.get_user(user_id)
        if not user:
            raise ValueError('User not found')
        if user.credits < amount:
            raise ValueError('Insufficient credits')

        new_balance = user.credits - amount
        await self.cosmos_service.update_user_credits(user_id, new_balance)

        transaction = CreditTransaction(
            id=str(uuid.uuid4()),
            user_id=user_id,
            amount=-amount,
            type='DEDUCTION',
            description=description,
            created_at=datetime.utcnow().isoformat()
        )
        await self.cosmos_service.create_transaction(transaction)

        return new_balance

    async def add_credits(self, user_id: str, amount: int, description: str, reference: Optional[str] = None) -> int:
        user = await self.cosmos_service.get_user(user_id)
        if not user:
            raise ValueError('User not found')

        new_balance = user.credits + amount
        await self.cosmos_service.update_user_credits(user_id, new_balance)

        transaction = CreditTransaction(
            id=str(uuid.uuid4()),
            user_id=user_id,
            amount=amount,
            type='PURCHASE',
            description=description,
            reference=reference,
            created_at=datetime.utcnow().isoformat()
        )
        await self.cosmos_service.create_transaction(transaction)

        return new_balance
    
    async def get_user_transactions(self, user_id: str) -> List[CreditTransaction]:
        """Get user's transaction history"""
        return await self.cosmos_service.get_user_transactions(user_id)